pageserver/src/page_service.rs

```diff
@@ -219,21 +213,55 @@ pub fn spawn_grpc(
         .build_v1()?;
     let server = server.add_service(reflection_service);
 
-    // Spawn server task.
-    let task_cancel = cancel.clone();
-    let task = COMPUTE_REQUEST_RUNTIME.spawn(task_mgr::exit_on_panic_or_error(
-        "grpc listener",
-        async move {
-            let result = server
-                .serve_with_incoming_shutdown(incoming, task_cancel.cancelled())
-                .await;
-            if result.is_ok() {
-                // TODO: revisit shutdown logic once page service is implemented.
-                gate.close().await;
-            }
-            result
-        },
-    ));
+    // Set up the TCP socket. We take a preconfigured TcpListener to bind the port early.
+    listener.set_nonblocking(true)?;
+    setsockopt(&listener, sockopt::KeepAlive, &GRPC_TCP_KEEPALIVE)?;
+    let listener = tokio::net::TcpListener::from_std(listener)?;
+
+    // Build the serve future.
+    let cancel_serve = cancel.clone();
+    let serve = async move {
+        // Accept TCP connections.
+        let tcp_conns = TcpListenerStream::new(listener).map(|result| {
+            let tcp_conn = result.inspect_err(|err| error!("TCP accept failed: {err}"))?;
+            tcp_conn.set_nodelay(GRPC_TCP_NODELAY).inspect_err(|err| {
+                error!("TCP nodelay failed: {err}");
+            })?;
+            Ok(tcp_conn)
+        });
+
+        if let Some(tls_config) = tls_config {
+            // If TLS is enabled, decrypt the TCP streams before passing them to the server.
+            let tls_acceptor = TlsAcceptor::from(tls_config);
+            let tls_conns = async_stream::stream! {
+                for await result in tcp_conns {
+                    match result {
+                        Ok(tcp_conn) => yield tls_acceptor
+                            .accept(tcp_conn)
+                            .await
```

Comment:
It's not recommended to implement TLS like that

The current flow of the execution will be:
```
TCP accept (1) -> TLS handshake (1) -> TCP accept (2) -> TLS handshake (2) -> ...
```
The problem is that it takes some time to perform TLS handshake (RTT), and during TLS handshake you can't accept other connection. Even worse, if the client hangs during TLS handshake, you won't be able to accept other connection till the TLS handshake is timed out.

The recommended way is to run a loop which only accepts TCP connections and then spawn a new tokio task to perform TLS handshake and execute the request.

That's what the example of tonic_rustls does:
https://github.com/hyperium/tonic/blob/689a86dbad107fa721714bb02e7f8e39264a377d/examples/src/tls_rustls/server.rs#L68